x/concentrated-liquidity/position.go

```diff
@@ -621,17 +610,54 @@ func (k Keeper) RemovePositionIdToLock(ctx sdk.Context, positionId, underlyingLo
 	store.Delete(lockIdKey)
 }
 
-// positionHasUnderlyingLockInState checks if a given positionId has a corresponding lock in state.
-func (k Keeper) positionHasUnderlyingLockInState(ctx sdk.Context, positionId uint64) (bool, error) {
+// PositionHasActiveUnderlyingLock is a non mutative method that checks if a given positionId has a corresponding lock in state.
+// If it has a lock in state, checks if that lock is still active.
+// If lock is still active, returns true.
+// If lock is no longer active, returns false.
+func (k Keeper) PositionHasActiveUnderlyingLock(ctx sdk.Context, positionId uint64) (hasActiveUnderlyingLock bool, lockId uint64, err error) {
 	// Get the lock ID for the position.
-	_, err := k.GetLockIdFromPositionId(ctx, positionId)
-	if err == nil || !errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
-		return true, nil
+	lockId, err = k.GetLockIdFromPositionId(ctx, positionId)
+	if errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
+		return false, 0, nil
+	} else if err != nil {
+		return false, 0, err
+	}
+
+	// Check if the underlying lock is mature.
+	lockIsMature, err := k.isLockMature(ctx, lockId)
+	if err != nil {
+		return true, 0, err
+	}
+	if lockIsMature {
+		return false, 0, nil
 	}
+	return true, lockId, nil
+}
+
+// PositionHasActiveUnderlyingLockAndUpdate is a mutative method that checks if a given positionId has a corresponding lock in state.
+// If it has a lock in state, checks if that lock is still active.
+// If lock is still active, returns true.
+// If lock is no longer active, removes the lock ID from the position ID to lock ID mapping and returns false.
+func (k Keeper) PositionHasActiveUnderlyingLockAndUpdate(ctx sdk.Context, positionId uint64) (hasActiveUnderlyingLock bool, lockId uint64, err error) {
+	// Get the lock ID for the position.
+	lockId, err = k.GetLockIdFromPositionId(ctx, positionId)
 	if errors.Is(err, types.PositionIdToLockNotFoundError{PositionId: positionId}) {
-		return false, nil
+		return false, 0, nil
+	} else if err != nil {
+		return false, 0, err
+	}
+
+	// Check if the underlying lock is mature.
+	lockIsMature, err := k.isLockMature(ctx, lockId)
+	if err != nil {
+		return true, 0, err
+	}
+	if lockIsMature {
```

Comment:
Can we use ` PositionHasActiveUnderlyingLock` here?

Reply:
As it currently is structured, no, because `PositionHasActiveUnderlyingLock` returns false, 0, nil when both the lockIsMature as well as if the PositionIdToLockNotFoundError is present. So there is no way to know what the previous lockId was.

Reply:
I don't follow

When lockIsMature as well as if the PositionIdToLockNotFoundError, there is no lock exists that we should be removing because 0 signifies the lack of lock.

What lock id and why do we need to know? What am I missing?

Reply:
```
func (k Keeper) positionHasActiveUnderlyingLockAndUpdate(ctx sdk.Context, positionId uint64) (hasActiveUnderlyingLock bool, lockId uint64, err error) {
	// Get the lock ID for the position.
	hasActiveUnderlyingLock, lockId, err = k.PositionHasActiveUnderlyingLock(ctx, positionId)
	// There is currently no way to determine if lock `hasInactiveUnderlyingLock`
	// The case we want to handle is when hasActiveUnderlyingLock is false
	// We would ne to run:
	// if !hasActiveUnderlyingLock, re-retrieve the GetLockIdFromPositionId, and then remove the connection here
	if lockIsMature {
		k.RemovePositionIdToLock(ctx, positionId, lockId)
		return false, 0, nil
	}
	return true, lockId, nil
}
```

Reply:
Yeah the approach you mentioned seems preferable to me because it reduces complex edge case handling, reducing code duplication and only requiring one additional call to get the lock id if `hasActiveUnderlyingLock`.

```
 if !hasActiveUnderlyingLock, re-retrieve the GetLockIdFromPositionId, and then remove the connection here"
```
Re-fetching seems fine. Please consider this change

Reply:
https://github.com/osmosis-labs/osmosis/pull/5042/commits/5821706c879fb2f2ad772a4b4da0c8333dabc45f